Dockerfile

```diff
@@ -38,20 +22,48 @@ RUN sha256sum /tmp/sccache.tar.gz | grep e5d03a9aa3b9fac7e490391bbe22d4f42c840d3
 RUN tar -xf /tmp/sccache.tar.gz
 RUN mv sccache-v*/sccache /usr/local/bin/sccache
 RUN chmod +x /usr/local/bin/sccache
-RUN rm -rf sccache-v*/ /tmp/sccache.tar.gz
 
 # Check sccache version
 RUN sccache --version
 
+# Add scripts
+ADD optimize.sh /usr/local/bin/optimize.sh
+RUN chmod +x /usr/local/bin/optimize.sh
+
+ADD optimize_workspace.sh /usr/local/bin/
+RUN chmod +x /usr/local/bin/optimize_workspace.sh
+ADD build_workspace.py /usr/local/bin/
+RUN chmod +x /usr/local/bin/build_workspace.py
+
+#
+# base-optimizer
+#
+FROM rust:1.54.0-alpine as base-optimizer
+
+# Being required for gcc linking
+RUN apk update && \
+  apk add --no-cache musl-dev
+
+# Setup Rust with Wasm support
+RUN rustup target add wasm32-unknown-unknown
+
+# Add wasm-opt
+COPY --from=builder /usr/local/bin/wasm-opt /usr/local/bin
+
+#
+# rust-optimizer
+#
+FROM base-optimizer as rust-optimizer
+
 # Use sccache. Users can override this variable to disable caching.
+COPY --from=builder /usr/local/bin/sccache /usr/local/bin
 ENV RUSTC_WRAPPER=sccache
 
 # Assume we mount the source code in /code
 WORKDIR /code
 
-# Add out script as entry point
-ADD optimize.sh /usr/local/bin/optimize.sh
-RUN chmod +x /usr/local/bin/optimize.sh
+# Add script as entry point
+COPY --from=builder /usr/local/bin/optimize.sh /usr/local/bin
```

Comment:
Is there an advantage for copying them from the builder instead of the source repo directly?

Reply:
It reduces the size of the images because each `RUN chmod ...` command adds a layer

Reply:
Nice. Thanks for the clarification.

Reply:
This probably also explains why my attempts to shrink the image by deleting the native x86_64 compiler target failed 💡

Reply:
Yep that's right! I'd also thought to try to remove that target, but alas

Reply:
But we could install rustup on a plain Alpine and only install the wasm target, couldn't we? I'm not sure if Rust is happy without a native target installed.

Reply:
We could, but it might be best to just build on top of the public rust image. For reference, this is the Dockerfile 😫 https://github.com/rust-lang/docker-rust/blob/master/Dockerfile-alpine.template